# Review Word/Doc vectorization

#### Sparse

One hot Encoding. Here each word has a representation. There is no Doc representatation. (but it can be deduced)


In [17]:
from sklearn.preprocessing import OneHotEncoder
import numpy as np
import string

docs = [
    "Portugal will win world cup !", 
    "will Cristiano Ronaldo win the world cup ?", 
    "Portugal is the best team in the world",
    ]


words = [w.translate(str.maketrans('', '', string.punctuation)).lower() for w in " ".join(docs).split()]
vocabulary = sorted(set(words))
encoder = OneHotEncoder(sparse_output=False)
one_hot_vectors = encoder.fit_transform(np.array(vocabulary).reshape(-1, 1))

word_to_onehot = {vocabulary[i]: one_hot_vectors[i]
                  for i in range(len(vocabulary))}


for word, vector in word_to_onehot.items():
    print(f"Word: {word}, One-Hot Encoding: {vector}")

Word: , One-Hot Encoding: [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
Word: best, One-Hot Encoding: [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
Word: cristiano, One-Hot Encoding: [0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
Word: cup, One-Hot Encoding: [0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
Word: in, One-Hot Encoding: [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
Word: is, One-Hot Encoding: [0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
Word: portugal, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
Word: ronaldo, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]
Word: team, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
Word: the, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]
Word: will, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
Word: win, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]
Word: world, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]


Bad of words - frequency only. (Doc has representation.)

In [24]:
from sklearn.feature_extraction.text import CountVectorizer

docs = [
    "Portugal will win world cup !", 
    "will Cristiano Ronaldo win the world cup ?", 
    "Portugal is the best team in the world",
    ]


vectorizer = CountVectorizer()
X = vectorizer.fit_transform(docs)

print(X.toarray())
print(vectorizer.get_feature_names_out())


[[0 0 1 0 0 1 0 0 0 1 1 1]
 [0 1 1 0 0 0 1 0 1 1 1 1]
 [1 0 0 1 1 1 0 1 2 0 0 1]]
['best' 'cristiano' 'cup' 'in' 'is' 'portugal' 'ronaldo' 'team' 'the'
 'will' 'win' 'world']


TF-IDF

- TF: Term Frequency, measures the frequency that the term appears in the document. how many times appear vs numver of total woords/tokens.
- IDF: Iverse Document Frequency, measures he importance of the word in the entire corpus. The more documents the word appears in the less important it is: total Number of docs (log(of that))

compared to the others this highlight the importannt words by reducing the weight of the common words. (like stop words)

In [26]:
from sklearn.feature_extraction.text import TfidfVectorizer

docs = [
    "Portugal will win world cup !", 
    "will Cristiano Ronaldo win the world cup ?", 
    "Portugal is the best team in the world",
    ]

tfidf_vectorizer = TfidfVectorizer()
X_tfidf = tfidf_vectorizer.fit_transform(docs)

print(X_tfidf.toarray())
print(tfidf_vectorizer.get_feature_names_out())

[[0.         0.         0.46609584 0.         0.         0.46609584
  0.         0.         0.         0.46609584 0.46609584 0.361965  ]
 [0.         0.46312056 0.35221512 0.         0.         0.
  0.46312056 0.         0.35221512 0.35221512 0.35221512 0.27352646]
 [0.37162591 0.         0.         0.37162591 0.37162591 0.28263108
  0.         0.37162591 0.56526217 0.         0.         0.21948825]]
['best' 'cristiano' 'cup' 'in' 'is' 'portugal' 'ronaldo' 'team' 'the'
 'will' 'win' 'world']


[[0.        0.         0.466095 0.         0.         0.46609584 0.         0.         0.         0.46609584 0.466095 0.361965  ]
[0.         0.46312056 0.352215 0.         0.         0.         0.46312056 0.         0.35221512 0.35221512 0.352215 0.27352646]
[0.37162591 0.         0.       0.37162591 0.37162591 0.28263108 0.         0.37162591 0.56526217 0.         0.       0.21948825]]
['best'     'cristiano''cup'     'in'       'is'       'portugal'  'ronaldo'  'team'   'the'      'will'    'win'     'world']

#### Dense

Word embeddings

## Dense Vectorization (Static vs. Contextual embeddings)
#### there are word vs sentnece embedding look up this later

**Dense vs. sparse.** This is about the *shape* of the vector, not the technique. TF-IDF is **sparse**: tens of thousands of dimensions (one per vocabulary term), almost all zero. A **dense vector** is the opposite — few dimensions (typically 100–1000), all filled with real numbers, with meaning distributed across all positions instead of one dimension per word.

**Dense ≠ transformers.** "Embeddings" are dense vectors that capture *meaning* (semantically similar texts land close together). Transformers are the most powerful way to produce them today, but not the only one — word embeddings predate transformers by years.

**The spectrum of dense approaches (lightest → heaviest):**

1. **Classic word embeddings — Word2Vec, GloVe, FastText.** No transformers. One dense vector per *word*; to classify a sentence you average its word vectors and feed that into a classic classifier. Fast, CPU-friendly. Limitations: non-contextual (same vector for a word regardless of meaning) and averaging loses word order.

2. **Sentence embeddings — Sentence-BERT (SBERT).** Transformers enter here. Models like `all-MiniLM-L6-v2` turn a whole sentence directly into a dense vector (e.g. 384 dims) that captures its meaning *in context*. Pre-trained, so inference is cheap (runs on CPU for our dataset size) — we only encode, we don't train the encoder.

3. **Full transformer fine-tuning — BERT, RoBERTa.** Train the transformer itself end-to-end for the task. Maximum performance, but needs a GPU, more data, more time, and more complexity.

**Chosen approach for this task: SBERT sentence embeddings + a classic classifier.**
- Fits the existing workflow: swap `TfidfVectorizer` for an SBERT encoder (query → 384-dim dense vector), then feed those vectors into `LogisticRegression` / `LinearSVC`. The "vectorizer → classifier" structure stays; only the first half changes.
- Best test of semantic value on Banking77: TF-IDF fails when two intents use *different words for the same idea* (synonyms) or when meaning is context-dependent — exactly where semantic embeddings help, since they place similar sentences close even with no shared words.

**Important compatibility note:** SBERT embeddings contain **negative values**, and `MultinomialNB` does **not** accept negative features (it assumes counts ≥ 0). So the multinomial Naive Bayes cannot be reused with dense vectors. Dense vectors pair naturally with classifiers that handle arbitrary real-valued features: `LogisticRegression`, `LinearSVC`, or `GaussianNB`. `LogisticRegression` is the classic, robust choice for embeddings.

**Dependency:** `pip install sentence-transformers`. First run downloads the model (a few hundred MB), then runs locally with no API or internet.

In [27]:
#TODO